### PR TITLE
Builder classes for service and endpoint

### LINKED ISSUE #836: Create `Builder` data classes for `Service` and `Endpoint`
Context : https://github.com/disneystreaming/smithy4s/pull/832#issuecomment-1446002920

Let's create `Endpoint.Builder` and `Service.Builder` data classes that can be easily access from `Endpoint` and `Service`, that would facilitate the manipulation and transformation of `Endpoint` and `Service`

They should follow the [Scala best practice around binary compatibility](https://docs.scala-lang.org/overviews/core/binary-compatibility-for-library-authors.html#changing-a-case-class-definition-in-a-backwards-compatible-manner).  

We could enhance the UX of using builders by having `mapX` methods for every element.

### PR DESCRIPTION
**Pull request by GaryAghedo**
Closes #836.

### CODE REVIEW

File: modules/core/src/smithy4s/Service.scala
Comment (kubukoz) [inline modules/core/src/smithy4s/Service.scala]: @GaryAghedo can you add methods for these? It'd be the usual, just like everything in the endpoint builder.

**Review by kubukoz — COMMENTED** [review]
Thanks for kicking this off! I took your code as a baseline and implemented the endpoint transformation function in the service mapper. I'm not sure it's fully correct, we'll have to get @Baccata's take on this, but I think we're onto something.

We'll need some tests before this can be merged, there are probably also other things...

File: modules/core/src/smithy4s/Service.scala
Comment (GaryAghedo) [reply modules/core/src/smithy4s/Service.scala]: nice!
Yes sure.. I will add the code now

File: modules/core/src/smithy4s/Endpoint.scala
Comment (kubukoz) [inline modules/core/src/smithy4s/Endpoint.scala]: **question:** should we allow using a schema of a different type (another `O`, in this case)? I'm thinking yes, because it's certainly possible.

However, when that's used in conjunction with a service builder, I think we'd be pretty limited here - not only because `PolyFunction5` forbids changing type parameters, but also I don't think we can change the operation type (which would be necessary to keep things like `def endpoint: (I, Endpoint)` working).

Comment (Baccata) [reply modules/core/src/smithy4s/Endpoint.scala]: I think there could eventually be two flavours of builders : 

1. One for the static usecase, that respects all the types, which probably means that the only possible transformations will be hints and labels 
2. One that would be more useful in the context of dynamic services, where we could amend both the types involved in operations, and the list of operations 

I believe this PR is only dealing with point 1 (which is totally fine). In this context, you shouldn't allow for different types.

File: modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala
Comment (kubukoz) [inline modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala:1]: **issue:** this should go to `core/test` - it has a dependency on `bootstrapped` so you should be able to move it pretty much directly

Comment (Baccata) [reply modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala:1]: actually it doesn't have a dependency and the test is right where it should be. Attempting to add a dependency from `core/Test` to `bootstrapped` created a cycle because of SBT's model which led the build to crash. 

`bootstrapped/Test` now holds all the tests that were previously in core/Test. Blame SBT.

Comment (kubukoz) [reply modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala:1]: oh. welp, I should've double-checked then :P

File: modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala
Comment (kubukoz) [inline modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala]: **suggestion:** 
```suggestion
        hints ++ Hints(Documentation("new endpoint"))
```

with `Documentation` coming from the `smithy.api` package

File: modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala
Comment (kubukoz) [inline modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala]: **nitpick:** `ShapeId.apply` -> `ShapeId`

File: modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala
Comment (kubukoz) [inline modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala]: **nitpick:** this will probably fail formatting

File: modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala
Comment (kubukoz) [inline modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala]: **suggestion:** you can access this particular endpoint with `smithy4s.example.WeatherOperation.GetForecast` rather tahn indexing

File: modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala
Comment (kubukoz) [inline modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala:48]: **thought:** this could have some interesting implicitations if someone does that on an endpoint that has errors: I wonder what happens if they use the "surfaced errors" way of writing services and raise an error that way...

File: modules/bootstrapped/test/src/smithy4s/builder/ServiceBuilderSpec.scala
Comment (kubukoz) [inline modules/bootstrapped/test/src/smithy4s/builder/ServiceBuilderSpec.scala]: **issue:** same suggestion as in the endpoint test, you can pass hints with less boilerplate

File: modules/bootstrapped/test/src/smithy4s/builder/ServiceBuilderSpec.scala
Comment (kubukoz) [inline modules/bootstrapped/test/src/smithy4s/builder/ServiceBuilderSpec.scala]: **suggestion:** 

```suggestion
          Hints(ShapeId.apply("smithy.api", "documentation2") -> Document.fromString("new Service")))
```

(there's an implicit conversion)

File: modules/bootstrapped/test/src/smithy4s/builder/ServiceBuilderSpec.scala
Comment (kubukoz) [inline modules/bootstrapped/test/src/smithy4s/builder/ServiceBuilderSpec.scala:10]: **nitpick:** I would've liked to see a service with multiple endpoints used in this test - allows for covering more ground in the tests (e.g. you see that endpoints aren't all replaced with the same thing)

File: modules/core/src/smithy4s/Endpoint.scala
Comment (kubukoz) [inline modules/core/src/smithy4s/Endpoint.scala:81]: **thought:** I didn't suggest adding the ability to customize `wrap` because, frankly, what can you do? other than side effects like printing the input, I don't see how else you might extend this. Just leaving this for other maintainers if they're wondering why `wrap` doesn't have builder functions.

File: modules/core/src/smithy4s/Service.scala
Comment (kubukoz) [inline modules/core/src/smithy4s/Service.scala]: **thought:** this transformation is kind of lazy (only happens when you call `.build`), perhaps we should document that in its Scaladoc.

also, perhaps we should only cache it inside `build`, otherwise if you call `mapEndpointEach` multiple times you get multiple layers of caching.

File: modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala
Comment (kubukoz) [reply modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala:48]: I think it's worth a test in here, at least.

Comment (kubukoz) [reply modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala:48]: just checked, the behavior is:

- if you remove an errorable and the service raises an error, it behaves just like on any other Throwable
- adding errors to an existing errorable, or creating custom Alts in the errorable otherwise, seems impossible (legally) because `E` isn't known: so you can't biject an arbitrary value into it

File: modules/core/src/smithy4s/Service.scala
Comment (Baccata) [reply modules/core/src/smithy4s/Service.scala]: NB : syncing with series `0.18`  will remove `unsafeCacheBy`, but it's okay because the exposure of ordinals should remove the need to cache.

File: modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala
Comment (GaryAghedo) [reply modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala]: https://github.com/disneystreaming/smithy4s/pull/1097/commits/748de5300df1a328d5c8eb09dcf1c613783e7b70

File: modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala
Comment (GaryAghedo) [reply modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala]: https://github.com/disneystreaming/smithy4s/pull/1097/commits/6346866132e5c7ce91ea50417345b6d6488ab705

File: modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala
Comment (GaryAghedo) [reply modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala]: https://github.com/disneystreaming/smithy4s/pull/1097/commits/6346866132e5c7ce91ea50417345b6d6488ab705

File: modules/bootstrapped/test/src/smithy4s/builder/ServiceBuilderSpec.scala
Comment (GaryAghedo) [reply modules/bootstrapped/test/src/smithy4s/builder/ServiceBuilderSpec.scala]: https://github.com/disneystreaming/smithy4s/pull/1097/commits/8ffcacfa0cc5a7e2446723e3efd0e8e165f8feb1

File: modules/bootstrapped/test/src/smithy4s/builder/ServiceBuilderSpec.scala
Comment (GaryAghedo) [reply modules/bootstrapped/test/src/smithy4s/builder/ServiceBuilderSpec.scala]: https://github.com/disneystreaming/smithy4s/pull/1097/commits/a01028fa70381c2d8b7db7a0bfb4741f1b306421

File: modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala
Comment (GaryAghedo) [reply modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala]: https://github.com/disneystreaming/smithy4s/pull/1097/commits/e0ebde221291632b48a0f58372ef5eeb8dbf600c

File: modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala
Comment (GaryAghedo) [reply modules/bootstrapped/test/src/smithy4s/builder/EndpointBuilderSpec.scala:48]: https://github.com/disneystreaming/smithy4s/pull/1097/commits/f31a91264de502e1a1e2b721bcd61c746f60b541

File: modules/core/src/smithy4s/Endpoint.scala
Comment (GaryAghedo) [reply modules/core/src/smithy4s/Endpoint.scala:81]: if customization of the wrap is not available, we must offer an alternative solution that involves providing an implementation. Its seems unavoidable in this case

Comment (kubukoz) [reply modules/core/src/smithy4s/Endpoint.scala:81]: I had a thought the other day, now I'm thinking `wrap` could actually be customizable on an individual endpoint level. so, `withWrap` and `mapWrap` would still be doable.

I think we can leave them out of this PR though, and make a separate ticket.

Comment (kubukoz) [reply modules/core/src/smithy4s/Endpoint.scala:81]: https://github.com/disneystreaming/smithy4s/issues/1152

File: modules/core/src/smithy4s/Service.scala
Comment (kubukoz) [inline modules/core/src/smithy4s/Service.scala:218]: **thought:** lost this comment somewhere, but I had some doubts about this method: it's lazy (`mapper` doesn't get called until you call `build`), which could be surprising to users. However, mapping the endpoints eagerly proved to be a bit difficult last time I tried to do it, so I don't have a great suggestion either.

Only one: let's document the lazy behavior: sth like

```scala
/**
  Maps each endpoint in the service with the given polymorphic function. Note: unlike other `map*` methods here, this one is lazy and non-memoized - the mapping is only performed when you call `build`, every time you call it.
```

File: modules/core/src/smithy4s/Service.scala
Comment (kubukoz) [inline modules/core/src/smithy4s/Service.scala]: **nitpick:** this could probably be a `val`

File: build.sbt
Comment (kubukoz) [inline build.sbt]: **nitpick:** I thought `bootstrapped` was the only module that generates code, but apparently not.

however, this should probably be scoped to `Test` instead of `Compile`, just like the other codegen-specific settings here.

**Review by kubukoz — APPROVED** [review]

File: modules/core/src/smithy4s/Service.scala
Comment (kubukoz) [reply modules/core/src/smithy4s/Service.scala]: it did, because we got rid of `def endpoint`. We'll defer to the default impl, unless you prefer otherwise.

File: modules/core/src/smithy4s/Service.scala
Comment (kubukoz) [reply modules/core/src/smithy4s/Service.scala:218]: this is outdated now, the transformation is no longer lazy.

**Comment by kubukoz** [comment]
@Baccata can you take another look?

**Review by Baccata — APPROVED** [review]
Thanks !

**Pull request merged**
